blockchain/blockchain.go

```diff
@@ -0,0 +1,182 @@
+//go:generate abigen --abi ../resources/Agent.abi --pkg blockchain --type Agent --out agent.go
+
+package blockchain
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"encoding/hex"
+	"encoding/json"
+	"github.com/coreos/bbolt"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/singnet/snet-daemon/config"
+	"github.com/singnet/snet-daemon/db"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+const (
+	JobPendingState    = "PENDING"
+	JobFundedState     = "FUNDED"
+	JobAddressHeader   = "snet-job-address"
+	JobSignatureHeader = "snet-job-signature"
+)
+
+type jobInfo struct {
+	jobAddressBytes   []byte
+	jobSignatureBytes []byte
+}
+
+var (
+	ethClient          *ethclient.Client
+	rawClient          *rpc.Client
+	agent              *Agent
+	privateKey         *ecdsa.PrivateKey
+	address            string
+	jobCompletionQueue chan *jobInfo
+)
+
+func init() {
+	if config.GetBool(config.BlockchainEnabledKey) {
+		// Setup ethereum client
+		if client, err := rpc.Dial(config.GetString(config.EthereumJsonRpcEndpointKey)); err != nil {
+			log.WithError(err).Panic("error creating rpc client")
+		} else {
+			rawClient = client
+			ethClient = ethclient.NewClient(rawClient)
+		}
+
+		// Setup agent
+		if a, err := NewAgent(common.HexToAddress(config.GetString(config.AgentContractAddressKey)), ethClient); err != nil {
+			log.WithError(err).Panic("error instantiating agent")
+		} else {
+			agent = a
+		}
+
+		// Setup identity
+		if privateKeyString := config.GetString(config.PrivateKeyKey); privateKeyString != "" {
+			if privKey, err := crypto.HexToECDSA(privateKeyString); err != nil {
+				log.WithError(err).Panic("error getting private key")
+			} else {
+				privateKey = privKey
+				address = crypto.PubkeyToAddress(privateKey.PublicKey).Hex()
+			}
+		} else if hdwalletMnemonic := config.GetString(config.HdwalletMnemonicKey); hdwalletMnemonic != "" {
+			if privKey, err := derivePrivateKey(hdwalletMnemonic, 44, 60, 0, 0, uint32(config.GetInt(config.HdwalletIndexKey))); err != nil {
+				log.WithError(err).Panic("error deriving private key")
+			} else {
+				privateKey = privKey
+				address = crypto.PubkeyToAddress(privateKey.PublicKey).Hex()
+			}
+		}
+
+		// Start event and job completion routines
+		jobCompletionQueue = make(chan *jobInfo, 1000)
+		go processJobCompletions()
```

Comment:
Doing stuff like this in `init()` is frowned upon. We should avoid spawning goroutines or really doing much of anything as a side effect of importing a package. Let's talk offline about this pattern.